carma_wm_ros2/include/carma_wm_ros2/WMListener.hpp

```diff
@@ -0,0 +1,153 @@
+   */
+  WMListener(
+    const rclcpp::NodeOptions& options,
+    rclcpp::node_interfaces::NodeBaseInterface::SharedPtr node_base,
+    rclcpp::node_interfaces::NodeLoggingInterface::SharedPtr node_logging,
+    rclcpp::node_interfaces::NodeTopicsInterface::SharedPtr node_topics,
+    bool multi_thread = false);
+
+  /*!
+   * \brief Returns a pointer to an intialized world model instance
+   *
+   * \return Const pointer to a world model object
+   */
+  WorldModelConstPtr getWorldModel();
+
+  /*!
+   * \brief Allows user to set a callback to be triggered when a map update is received
+   *        NOTE: If operating in multi-threaded mode the world model will remain locked until the user function
+   * completes.
+   *
+   * \param callback A callback function that will be triggered after the world model receives a new map update
+   */
+  void setMapCallback(std::function<void()> callback);
+
+  /*!
+   * \brief Allows user to set a callback to be triggered when a route update is received
+   *        NOTE: If operating in multi-threaded mode the world model will remain locked until the user function
+   * completes.
+   *
+   * \param callback A callback function that will be triggered after the world model is updated with a new route
+   */
+  void setRouteCallback(std::function<void()> callback);
+
+  /*!
+   * \brief Returns a unique_lock which can be used to lock world model updates until the user finishes a desired
+   * operation. This function should be used when multiple queries are needed and this object is operating in
+   * multi-threaded mode
+   *
+   * \param pre_locked If true the returned lock will already be locked. If false the lock will be deferred with
+   * std::defer_lock Default is true
+   *
+   * \return std::unique_lock for thread safe access to world model data
+   */
+  std::unique_lock<std::mutex> getLock(bool pre_locked = true);
+
+  /*!
+   * \brief Allows user to set a callback to be triggered when a route update is received
+   *        NOTE: If operating in multi-threaded mode the world model will remain locked until the user function
+   * completes.
+   *
+   * \param config_lim A function that populate the configurable speed limit value after the world model is updated with a new route
+   */
+  void setConfigSpeedLimit(double config_lim) const;
+
+  /*!
+   * \brief Use to allow updates to occur even if they invalidate the current route.
+   *        This is only meant to be used by components which generate the route
+   */  
+  void enableUpdatesWithoutRouteWL();
+
+  /*!
+   * \brief Returns true if a map update has been processed which requires rerouting.
+   *        This method is meant to be used by routing components. 
+   * 
+   * \return True is rerouting is needed
+   */ 
+  bool checkIfReRoutingNeededWL();
+
+  ////
+    // Overrides
+    ////
+  carma_ros2_utils::CallbackReturn handle_on_configure(const rclcpp_lifecycle::State &);
```

Comment:
Since we don't want this to be separate from the core node we should probably just do all the configuration on construction with the expectation that the user node includes the WMListener() call in their own handle_on_configure method.